src/bin/vip-dev-env-shell.js

```diff
@@ -0,0 +1,101 @@
+
+/**
+ * Internal dependencies
+ */
+import { trackEvent } from '../lib/tracker';
+import command from '../lib/cli/command';
+import { getEnvironmentPath } from '../lib/dev-environment/dev-environment-core';
+import { DEV_ENVIRONMENT_FULL_COMMAND } from '../lib/constants/dev-environment';
+import { getEnvTrackingInfo, validateDependencies, getEnvironmentName, handleCLIException } from '../lib/dev-environment/dev-environment-cli';
+import { bootstrapLando, landoShell } from '../lib/dev-environment/dev-environment-lando';
+
+const debug = debugLib( '@automattic/vip:bin:dev-environment' );
+
+const userMap = {
+	devtools: 'www-data',
+	nginx: 'www-data',
+	php: 'www-data',
+	database: 'mysql',
+	memcached: 'memcache',
+	wordpress: 'www-data',
+	'vip-mu-plugins': 'www-data',
+	'demo-app-code': 'www-data',
+	elasticsearch: 'elasticsearch',
+	phpmyadmin: 'www-data',
+	mailhog: 'mailhog',
+};
+
+const examples = [
+	{
+		usage: `${ DEV_ENVIRONMENT_FULL_COMMAND } shell`,
+		description: 'Spawns a shell in the dev environment',
+	},
+	{
+		usage: `${ DEV_ENVIRONMENT_FULL_COMMAND } shell -r`,
+		description: 'Spawns a shell in the dev environment under root user',
+	},
+	{
+		usage: `${ DEV_ENVIRONMENT_FULL_COMMAND } shell -- ls -lha`,
+		description: 'Runs `ls -lha` command in the shell in the dev environment',
+	},
+	{
+		usage: `${ DEV_ENVIRONMENT_FULL_COMMAND } shell -S database -- ls -lha`,
+		description: 'Runs `ls -lha` command in the shell of the database service in the dev environment',
+	},
+];
+
+function getCommand( args: string[] ): string[] {
+	const splitterIdx = process.argv.findIndex( argument => '--' === argument );
+	if ( args.length > 0 && splitterIdx === -1 ) {
+		throw new Error( 'Please provide "--" argument to separate arguments for "vip" and command to be executed (see "--help" for examples)' );
+	}
+
+	let cmd: string[] = [];
+	if ( splitterIdx !== -1 && splitterIdx + 1 < process.argv.length ) {
+		cmd = process.argv.slice( splitterIdx + 1 );
+	}
+
+	return cmd;
+}
+
+command( { wildcardCommand: true } )
+	.option( 'slug', 'Custom name of the dev environment' )
+	.option( 'root', 'Spawn a root shell' )
+	.option( 'service', 'Spawn a shell in a specific service (php if omitted)' )
+	.examples( examples )
+	.argv( process.argv, async ( args, opt ) => {
+		const slug = await getEnvironmentName( opt );
+
+		const lando = await bootstrapLando();
+		await validateDependencies( lando, '', true );
+
+		const trackingInfo = getEnvTrackingInfo( slug );
+		await trackEvent( 'dev_env_shell_command_execute', trackingInfo );
+
+		debug( 'Args: ', args, 'Options: ', opt );
+
+		const isRoot = !! opt.root;
+		const service = opt.service || 'php';
+		const user = isRoot ? 'root' : userMap[ service ] || 'www-data';
+		const cmd = getCommand( args );
+		try {
+			await landoShell( lando, getEnvironmentPath( slug ), service, user, cmd );
+			await trackEvent( 'dev_env_shell_command_success', trackingInfo );
+		} catch ( error ) {
```

Comment:
So this block is not entirely clear to me:

1) Where does error.hide come from?
2) Why do we report an error as success on #L98?

Reply:
1. From Lando. Lando uses that flag to distinguish between normal errors and non-zero exit code from commands.
2. Because we don't care if a command in the shell failed.

Reply:
Ok can we add that reasoning to a comment above the block as to not confuse the next person looking at this code?

Reply:
Updated the comment